cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/MetricsAdminSubscriberService.java

```diff
@@ -0,0 +1,155 @@
+
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+/**
+ * A TMS subscriber service for consuming metrics administrative messages and performance admin operations on the
+ * metrics system.
+ */
+public class MetricsAdminSubscriberService extends AbstractMessagingPollingService<MetricsAdminMessage> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(MetricsAdminSubscriberService.class);
+  private static final Gson GSON = new Gson();
+
+  // Number of messages to fetch per batch
+  private static final int FETCH_SIZE = 100;
+
+  private final MessagingContext messagingContext;
+  private final MetricDatasetFactory metricDatasetFactory;
+  private final MetricStore metricStore;
+  private MetricsConsumerMetaTable metaTable;
+
+  @Inject
+  MetricsAdminSubscriberService(CConfiguration cConf, MetricsCollectionService metricsCollectionService,
+                                MessagingService messagingService, MetricDatasetFactory metricDatasetFactory,
+                                MetricStore metricStore) {
+    super(NamespaceId.SYSTEM.topic(cConf.get(Constants.Metrics.ADMIN_TOPIC)),
+          metricsCollectionService.getContext(ImmutableMap.of(
+            Constants.Metrics.Tag.COMPONENT, Constants.Service.METRICS,
+            Constants.Metrics.Tag.NAMESPACE, NamespaceId.SYSTEM.getNamespace(),
+            Constants.Metrics.Tag.TOPIC, cConf.get(Constants.Metrics.ADMIN_TOPIC),
+            Constants.Metrics.Tag.CONSUMER, "metrics.admin"
+          )),
+          FETCH_SIZE, cConf.getLong(Constants.Metrics.ADMIN_POLL_DELAY_MILLIS),
+          RetryStrategies.fromConfiguration(cConf, "system.metrics."));
+
+    this.messagingContext = new MultiThreadMessagingContext(messagingService);
+    this.metricDatasetFactory = metricDatasetFactory;
+    this.metricStore = metricStore;
+  }
+
+  @Override
+  protected MessagingContext getMessagingContext() {
+    return messagingContext;
+  }
+
+  @Override
+  protected MetricsAdminMessage decodeMessage(Message message) {
+    return GSON.fromJson(message.getPayloadAsString(), MetricsAdminMessage.class);
+  }
+
+  @Nullable
+  @Override
+  protected String loadMessageId() {
+    MetricsConsumerMetaTable metaTable = getMetaTable();
+    TopicProcessMeta meta = metaTable.getTopicProcessMeta(new TopicIdMetaKey(getTopicId()));
+    if (meta == null) {
+      return null;
+    }
+    return Bytes.toString(meta.getMessageId());
+  }
+
+  @Nullable
+  @Override
+  protected String processMessages(Iterator<ImmutablePair<String, MetricsAdminMessage>> messages) throws Exception {
+    MetricsConsumerMetaTable metaTable = getMetaTable();
+
+    String messageId = null;
+    while (messages.hasNext()) {
+      ImmutablePair<String, MetricsAdminMessage> messagePair = messages.next();
+      MetricsAdminMessage message = messagePair.getSecond();
+
+      switch (message.getType()) {
+        case DELETE:
```

Comment:
can this take a long time and fail the TMS transaction? How would that be handled?

Reply:
There is no transaction involved. This class extends from `AbstractMessagingPollingService` which doesn't use transaction.

Reply:
But, it is a good point that it might timeout the connection when it try to fetch the next message, causing unnecessary logs (functionally still be correct). Moving to a separate queue to make the experience better.